### DISCUSSION TITLE
Setup instructions

### DISCUSSION THREAD

**Discussion by paulng30** [LoH-lu/netbox-nmap-scan#4] [Q&A] [answered]
Hi, 

I apologize I'm not well verse with Linux. Could you please provide instructions how to set this up from start?  Where can we place these files once they are downloaded?  Thank you very much!

**Comment by LoH-lu** [comment, marked as answer]
I will update the wiki to be more clear.
You can just create a folder in your home directory, it's enough to start with.

**Reply by paulng30** [reply]
Thank you so much!

**Reply by paulng30** [reply]
Hi, 

First, I want to say thank you for creating these scripts.  In my test runs, I found the scan script put everything in Global VRF.  How can I tell it assign certain prefix to certain VRF?

**Reply by LoH-lu** [reply]
Thanks for the feedback, can you try that version ?
https://github.com/LoH-lu/netbox-nmap-scan/releases/tag/0.1.1

Let me know if it work fine, it should update existing scanned IP already pushed to Netbox.
It will retrieve and reuse the same VRF than the one assigned to your prefixes

**Reply by paulng30** [reply]
Hi,

Thank you for the added VRF code.  I tested the new version.  I seems to work, except the 'netbox_push.py' threw this error at the end.  It was able to input all of the IP addresses into Netbox and throw this error.

Traceback (most recent call last):
  File "/opt/discovery/netbox_push2.py", line 88, in <module>
    write_data_to_netbox(url, token, 'ipam_addresses.csv')
  File "/opt/discovery/netbox_push2.py", line 80, in write_data_to_netbox
    future.result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 458, in result
    return self.__get_result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/discovery/netbox_push2.py", line 38, in process_row
    netbox.ipam.ip_addresses.create(
  File "/usr/local/lib/python3.10/dist-packages/pynetbox/core/endpoint.py", line 415, in create
    ).post(args[0] if args else kwargs)
  File "/usr/local/lib/python3.10/dist-packages/pynetbox/core/query.py", line 368, in post
    return self._make_call(verb="post", data=data)
  File "/usr/local/lib/python3.10/dist-packages/pynetbox/core/query.py", line 258, in _make_call
    raise RequestError(req)
pynetbox.core.query.RequestError: The request failed with code 400 Bad Request: {'vrf': ["Related object not found using the provided attributes: {'name':                                     ''}"]}

**Reply by paulng30** [reply]
Just an update, I ran it again with all prefixes set to 'auto scan'.  Now, I don't see the error message.  Must have been a data entry that the script didn't like.

**Reply by LoH-lu** [reply]
Hey thanks a lot for the feedback, it seems it was trying to push an empty name for the vrf, it was causing the issue.
Probably old data done on the first scan you did without the vrf part.
Let me know if it happen again.

**Reply by paulng30** [reply]
Hi LoH-lu,

Scripts work good, except I noticed whenever scan runs, it causes the end-devices to go offline for couple of seconds.  I watched the switch ports of the end-device while scanning, they show "INCOMPLETE".  Not sure if this is because the level and nmap scan that we use.  I wonder if we can tune it down a little bit.  Like only nmap scan on one particular port 443.  What's your thoughts?

**Reply by LoH-lu** [reply]
Hey,

This is the nmap command being run "nmap -sn -R -T3 --min-parallelism 10"

Ping Scan (-sn): Checks which hosts are up by sending ping requests, without scanning for open ports.
DNS Resolution (-R): Performs reverse DNS resolution on all target IPs, even those that do not respond to pings.
Normal Timing (-T3): Uses the default timing template, balancing speed and reliability.
Parallel Probes (--min-parallelism 10): Ensures at least 10 probes are active in parallel, enhancing scan speed.

I'm surprised to be honest as this nmap is not doing much.
Are all end devices going down or only specific one/brand ?

**Reply by paulng30** [reply]
hi,

Thanks for the detailed information on the nmap command.  I agreed this shouldn't cause any issues.  I'll run some more tests and keep you posted.  I have just a handful of devices in this one subnet that I'm testing, couple of Cisco switches, some Dell laptops.

I'm curious, would it be possible for me to exclude some prefixes in the 'netbox_retrieve.py' script when it writes out to csv file?  Reason is I have two locations and I would like to run these scripts in both locations and send back data to one Netbox, if all possible.

Thanks again!

**Reply by paulng30** [reply]
Found the cause of that issue and corrected it.  There was a loop on that test subnet.

**Reply by LoH-lu** [reply]
> hi,
> 
> Thanks for the detailed information on the nmap command. I agreed this shouldn't cause any issues. I'll run some more tests and keep you posted. I have just a handful of devices in this one subnet that I'm testing, couple of Cisco switches, some Dell laptops.
> 
> I'm curious, would it be possible for me to exclude some prefixes in the 'netbox_retrieve.py' script when it writes out to csv file? Reason is I have two locations and I would like to run these scripts in both locations and send back data to one Netbox, if all possible.
> 
> Thanks again!

Sorry, missed that message, i'll review that if you are still interested!

**Comment by lazirus33** [comment]
Hello, First I want to echo what the others have said and seriously thank you for creating these scripts. Total life saver, lol. I am getting an error when I get to the step for nmap_compare.py.  I keep getting the below error, It looks like it is referencing a "results" folder. However I do not see that in the folder tree. I have tried creating the folder, which did not help. Any advice would be appreciated, thanks! 

Traceback (most recent call last):
  File "/home/skywestne/projects/netbox-nmap-scan/nmap_compare.py", line 37, in <module>
    latest_files = get_latest_files(directory)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/skywestne/projects/netbox-nmap-scan/nmap_compare.py", line 29, in get_latest_files
    files = [f for f in os.listdir(directory) if f.endswith('.csv')]
                        ^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'results/'
skywestne@netbox:~/projects/netbox-nmap-scan$ ls

**Reply by LoH-lu** [reply]
Hey, thanks a lot for the feedback!
Have you run the nmap_scan_multi_dns.py before ?
Or have you got any error while running it ?

**Reply by lazirus33** [reply]
Yeah, I was going through the steps, I ran the nmap_scan_multi_dns.py, did not get any errors. I waited awhile to see if nmap was still running. Then I run the namp_compare and get the error

**Reply by LoH-lu** [reply]
I've pushed a fix so it check on which folder you are running the script before trying check/create the folder, can you double check everything runs fine now ?
https://github.com/LoH-lu/netbox-nmap-scan/releases/tag/0.2.1

**Reply by lazirus33** [reply]
Im still getting the error when running the nmap_compare. Does  nmap_scan_multi_dns.py create a results folder inside the current dir when it completes the nmap scan? Its weird that im not getting any errors when I run the scan script

**Reply by LoH-lu** [reply]
Yes it should create the folder during nmap_scan_multi_dns.py
Do you have a big network to scan ?

Because at the moment, it will create the result folder only if there is result file to put there.
So either the scan script has not finished yet, or the server you are running the script on can't reach any of your network.

I will add more logger to catch any errors like these.

**Reply by lazirus33** [reply]
Thank you for working with me on this. I have tagged all of the /8 and /16 to not scan. so it should only be scanning a bunch of /24's.  Attached is the output for the ipam.prefixes if that helps
[ipam_prefixes.txt](https://github.com/user-attachments/files/16071467/ipam_prefixes.txt)

**Reply by LoH-lu** [reply]
Ok I see now, here's the issue:
![image](https://github.com/LoH-lu/netbox-nmap-scan/assets/139378145/b616de9b-cfa2-41ca-9561-78a85c5ca18a)

You are not using the active status ? only container ?

**Reply by lazirus33** [reply]
Im completely new to Netbox, So I was following a walkthrough on the proper steps with setting up the ipam portion. They were saying to set your prefixes as container. I probably did it wrong, lol. Sorry, im going to go back through and change it. Ill keep you posted, thanks again!

**Reply by LoH-lu** [reply]
You can keep your /8, /16 as container (supernet)
Personnally we use Active as the end prefix, meaning there will be not other prefixes bellow

But tell me what you prefer, i can do a fix just for your case :)

**Reply by lazirus33** [reply]
That actually makes more sense to me, I had it that way to begin with and then changed it after the walkthrough, lol. Testing it now

**Reply by lazirus33** [reply]
You are awesome!! That worked. Thank you so much for the assistance with this. Great work!

**Comment by MROSSFTEK** [comment]
i just pulled your scripts down and ran that multi DNS file and its asking for a csv.. did i miss in the instructions where it tells you how to make that csv?

**Reply by LoH-lu** [reply]
Hey, yup you'll need to follow this:
https://github.com/LoH-lu/netbox-nmap-scan/wiki/How-to-run-%3F
The csv is created after you run netbox_retrieve.py (don't forget to fill out var.ini with your own Netbox details)

**Comment by luis15pt** [comment]
Similar issue my ipam_prefixes.csv is not filled? 
Prefix,VRF,Status,Tags,Tenant

Am i missing something from netbox?
![image](https://github.com/LoH-lu/netbox-nmap-scan/assets/100942871/fb8f73f5-76ce-45dc-9eda-2e499d638c89)

**Reply by LoH-lu** [reply]
Hey, run the netbox_retrieve.py and see if the prefixes csv file is filled up or not.
Once you run the scan and the scan is finished, the prefixes that have been scanned get removed from the prefixes csv file.
Let me know if that help

**Comment by luis15pt** [comment]
Found the issue the script looks for IPAM -> Prefixes that was empty

![image](https://github.com/LoH-lu/netbox-nmap-scan/assets/100942871/e6c90484-64ab-4fdd-86c3-3f34e64b8950)

**Reply by LoH-lu** [reply]
Ah yes of course! Good catch ;)

**Comment by luis15pt** [comment]
Also documentation doesn't mention that it needs nmap installed,

`sudo apt install nmap`

**Reply by LoH-lu** [reply]
Thanks! Added to the wiki
